sklearn/neighbors/base.py

```diff
@@ -103,6 +104,192 @@ def _get_weights(dist, weights):
+    copied = graph.format != 'csr'
+    graph = check_array(graph, accept_sparse='csr')
+    check_non_negative(graph, whom="precomputed distance matrix.")
+
+    if not _is_sorted_by_data(graph):
+        warnings.warn('Precomputed sparse input was not sorted by data.',
+                      EfficiencyWarning)
+        if not copied:
+            graph = graph.copy()
+
+        # if each sample has the same number of provided neighbors
+        row_nnz = np.diff(graph.indptr)
+        if row_nnz.max() == row_nnz.min():
+            n_samples = graph.shape[0]
+            distances = graph.data.reshape(n_samples, -1)
+
+            order = np.argsort(distances, kind='mergesort')
+            order += np.arange(n_samples)[:, None] * row_nnz[0]
+            order = order.ravel()
+            graph.data = graph.data[order]
+            graph.indices = graph.indices[order]
+
+        else:
+            for start, stop in zip(graph.indptr, graph.indptr[1:]):
+                order = np.argsort(graph.data[start:stop], kind='mergesort')
+                graph.data[start:stop] = graph.data[start:stop][order]
+                graph.indices[start:stop] = graph.indices[start:stop][order]
+    return graph
+
+
+def _kneighbors_from_graph(graph, n_neighbors, return_distance):
+    """Decompose a nearest neighbors sparse graph into distances and indices
+
+    Parameters
+    ----------
+    graph : CSR sparse matrix, shape (n_samples, n_samples)
+        Neighbors graph as given by kneighbors_graph or radius_neighbors_graph
+
+    n_neighbors : int
+        Number of neighbors required for each sample.
+
+    return_distance : boolean
+        If False, distances will not be returned
+
+    Returns
+    -------
+    neigh_dist : array, shape (n_samples, n_neighbors)
+        Distances to nearest neighbors. Only present if return_distance=True.
+
+    neigh_ind : array, shape (n_samples, n_neighbors)
+        Indices of nearest neighbors.
+    """
+    n_samples = graph.shape[0]
+    assert graph.format == 'csr'
+
+    # number of neighbors by samples
+    row_nnz = np.diff(graph.indptr)
+    row_nnz_min = row_nnz.min()
+    if n_neighbors is not None and row_nnz_min < n_neighbors:
+        raise ValueError(
+            '%d neighbors per samples are required, but some samples have only'
+            ' %d neighbors in precomputed graph matrix. Decrease number of '
+            'neighbors used or recompute the graph with more neighbors.'
+            % (n_neighbors, row_nnz_min))
+
+    # if each sample has the same number of provided neighbors
+    if row_nnz.max() == row_nnz_min:
+
+        def extract(a):
+            return a.reshape(n_samples, -1)[:, :n_neighbors]
+
+    else:
+        idx = np.tile(np.arange(n_neighbors), n_samples)
+        idx = idx.reshape(n_samples, n_neighbors)
+        idx += graph.indptr[:-1, np.newaxis]
+
+        def extract(a):
+            return a.take(idx, mode='clip').reshape(n_samples, n_neighbors)
```

Comment:
Having two code paths define two different extract functions is a little strange. What do you think of:

```py
def extract(a):
    # if each sample has the same number of provided neighbors
    if row_nnz.max() == row_nnz_min:
        return a.reshape(n_samples, -1)[:, :n_neighbors]
	
    idx = np.tile(np.arange(n_neighbors), n_samples)
    idx = idx.reshape(n_samples, n_neighbors)
    idx += graph.indptr[:-1, np.newaxis]
    return a.take(idx, mode='clip').reshape(n_samples, n_neighbors)
```